core/src/saros/net/xmpp/contact/AddContactHelper.java

```diff
@@ -0,0 +1,148 @@
+ * XMPP Roster.
+ *
+ * <p>TODO This code has space for improvements, but probably update smack first..
+ */
+class AddContactHelper {
+  private static final Logger log = Logger.getLogger(AddContactHelper.class);
+
+  private static final class DialogContent {
+    private final String dialogTitle;
+    private final String dialogMessage;
+    private final String invocationTargetExceptionMessage;
+
+    DialogContent(
+        String dialogTitle, String dialogMessage, String invocationTargetExceptionMessage) {
+      this.dialogTitle = dialogTitle;
+      this.dialogMessage = dialogMessage;
+      this.invocationTargetExceptionMessage = invocationTargetExceptionMessage;
+    }
+  }
+
+  private AddContactHelper() {
+    // Helper Class
+  }
+
+  /**
+   * @param connectionService
+   * @param jid
+   * @param nickname
+   * @param questionDialog
+   * @throws SarosCancellationException If information about JID can not be found / retrieved from
+   *     Server and User canceled further trying, or Smack experienced an error.
+   */
+  static void addToRoster(
+      XMPPConnectionService connectionService,
+      JID jid,
+      String nickname,
+      BiPredicate<String, String> questionDialog)
+      throws SarosCancellationException {
+    Connection connection =
+        Objects.requireNonNull(connectionService.getConnection(), "connection is null");
+    Objects.requireNonNull(jid, "jid is null");
+
+    try {
+      if (!XMPPUtils.isJIDonServer(connection, jid, SarosConstants.RESOURCE)) {
+        if (!questionDialog.test(
+            "Contact Unknown",
+            "You entered a valid XMPP server.\n\n"
+                + "Unfortunately your entered JID is unknown to the server.\n"
+                + "Please make sure you spelled the JID correctly.\n\n"
+                + "Do you want to add the contact anyway?")) {
+          throw new SarosCancellationException("Please make sure you spelled the JID correctly.");
+        }
+
+        log.debug(
+            "The contact "
+                + jid
+                + " couldn't be found on the server."
+                + " The user chose to add it anyway.");
+      }
+    } catch (XMPPException e) {
+      DialogContent dialogContent = getDialogContent(e);
+
+      if (!questionDialog.test(dialogContent.dialogTitle, dialogContent.dialogMessage))
+        throw new SarosCancellationException(dialogContent.invocationTargetExceptionMessage);
+
+      log.warn(
+          "Problem while adding a contact.CancellationException User decided to add contact anyway. Problem: "
+              + e.getMessage());
+    }
+
+    try {
+      connection.getRoster().createEntry(jid.getBase(), nickname, null);
+    } catch (XMPPException e) {
+      log.error("Problem while adding a contact.", e);
+      throw new SarosCancellationException("Problem while adding a contact:" + e.getMessage());
+    }
+  }
+
+  private static DialogContent getDialogContent(XMPPException e) {
```

Comment:
Generating humanreadable error messages from error codes should be either part of the UI logic or some business logic in upper tiers and not in one of the lowest layer like the network. The whole code originally resided in the XMPPUtils class.

Reply:
I think this is the aspect we already talked about on other PRs: What do we want to do with IDE agnostic logic that provides information for the UI and is not pure core logic? This also contains the question: Which logic has to be the core?

Without blocking this PR we should answer these questions. I see the following approaches in order to separate the code:
1. separating the logic just in separate classes
1.  a separate package, the code is located in core
2. separating the core and core.ui in separate modules (req. JDK 9)
3.  a separate project with the ui code

I prefer approach 3 (or 2).
@stefaus  which approach do you prefer or do you have a totally different idea?